src/aggregators/coevolve.jl

```diff
@@ -0,0 +1,206 @@
+        rates = Vector{RateWrapper}()
+        lrates = Vector{RateWrapper}()
+        urates = Vector{RateWrapper}()
+        Ls = Vector{RateWrapper}()
+    end
+    cur_rates = nothing
+    sum_rate = nothing
+    next_jump = 0
+    next_jump_time = typemax(t)
+    CoevolveJumpAggregation(next_jump, next_jump_time, end_time, cur_rates, sum_rate,
+                            ma_jumps, rates, affects!, save_positions, rng;
+                            u = u,
+                            dep_graph = dep_graph,
+                            lrates = lrates, urates = urates, Ls = Ls)
+end
+
+# set up a new simulation and calculate the first jump / jump time
+function initialize!(p::CoevolveJumpAggregation, integrator, u, params, t)
+    p.end_time = integrator.sol.prob.tspan[2]
+    fill_rates_and_get_times!(p, u, params, t)
+    generate_jumps!(p, integrator, u, params, t)
+    nothing
+end
+
+# execute one jump, changing the system state
+function execute_jumps!(p::CoevolveJumpAggregation, integrator, u, params, t)
+    # execute jump
+    u = update_state!(p, integrator, u, params, t)
+    # update current jump rates and times
+    update_dependent_rates!(p, u, params, t)
+    nothing
+end
+
+# calculate the next jump / jump time
+function generate_jumps!(p::CoevolveJumpAggregation, integrator, u, params, t)
+    p.next_jump_time, p.next_jump = top_with_handle(p.pq)
+    nothing
+end
+
+@inline function update_state!(p::CoevolveJumpAggregation, integrator, u, params, t)
+    @unpack ma_jumps, next_jump = p
+    num_majumps = get_num_majumps(ma_jumps)
+    if next_jump <= num_majumps
+        if u isa SVector
+            integrator.u = executerx(u, next_jump, ma_jumps)
+        else
+            @inbounds executerx!(u, next_jump, ma_jumps)
+        end
+    else
+        idx = next_jump - num_majumps
+        @inbounds p.affects![idx](integrator)
+    end
+    p.prev_jump = next_jump
+    return integrator.u
+end
+
+######################## SSA specific helper routines ########################
+function update_dependent_rates!(p::CoevolveJumpAggregation, u, params, t)
+    @inbounds deps = p.dep_gr[p.next_jump]
+    @unpack end_time, pq = p
+    for (ix, i) in enumerate(deps)
+        ti = next_time(p, u, params, t, i, end_time)
+        update!(pq, i, ti)
+    end
+    nothing
+end
+
+function get_rates(p::CoevolveJumpAggregation, i, u)
+    ma_jumps = p.ma_jumps
+    num_majumps = get_num_majumps(ma_jumps)
+    if i <= num_majumps
+        _rate = evalrxrate(u, i, ma_jumps)
+        rate(u, p, t) = _rate
+        lrate = rate
+        urate = rate
+        L(u, p, t) = typemax(t)
+    else
+        idx = i - num_majumps
+        rate, lrate, urate, L = p.rates[idx], p.lrates[idx], p.urates[idx], p.Ls[idx]
+    end
+    return rate, lrate, urate, L
+end
+
+function next_time(p::CoevolveJumpAggregation{T}, u, params, t, i, tstop::T) where {T}
+    @unpack rng = p
+    rate, lrate, urate, L = get_rates(p, i, u)
+    while t < tstop
```

Comment:
I would rather handle all jumps with the same algorithm. 

I think that this module is not yet designed to make these type of of optimizations for the user. In ideal world, the user would not need to choose an aggregator and the library would 
decide for itself the best one. 

Since we are not there yet, I would argue that the current API promises the user that the same aggregator will be used for all discrete jumps in the problem. We should then honor this promise. Therefore, we should not surprise the user by silently calling aggregator B when aggregator A was invoked.

I think the penalty for using `Coevolve` instead of `NRM` for a problem involving only `MassActionJump`s and `ConstantRateJump`s is fixed and small enough that it should not make a huge difference for a naive user. As for an advanced user, she/he would know to call `NRM` when dealing with such problems.

With that being said, I think that `Coevolve` should have a performance very close to `NRM` in the cases involving only `MassActionJump`s and/or `ConstantRateJump`s  as the logic in the `next_time` function of `Coevolve` should reduce to the one in `NRM`. 

It seems like the penalty might be imposed from the `get_rates` function, but I'm not sure yet.

Reply:
I think you are going to have performance issues due to generating anonymous functions inside the hot loop (i.e. in `get_rates`). Generally creating functions is going to have an overhead, and moreover capturing a variable (`_rate`) can also have performance issues in Julia. Since there is no real reason to do this please take it out.

If you don't want to reuse `NRM`s time updates then just modify `next_time` to check if `i < (Number of mass action jumps + constant rate jumps)` and simply explicitly get the new rate via `evalrxrate` and/or the appropriate `rates[idx]`, and sample a new time for them. 

I suspect just making these changes will give performance quite close to `NRM` for systems with only `MassActionJump`s and `ConstantRateJump`s.